test/var_compatibility_test/effr_calc.F90

```diff
@@ -15,17 +15,19 @@ module effr_calc
    !> \section arg_table_effr_calc_run  Argument Table
    !! \htmlinclude arg_table_effr_calc_run.html
    !!
-   subroutine effr_calc_run(ncol, nlev, effrr_in, effrg_in, effrl_inout, &
-                            effri_out, effrs_inout, has_graupel, errmsg, errflg)
+   subroutine effr_calc_run(ncol, nlev, effrr_in, effrg_in, ncg_in, nci_out, &
+                            effrl_inout, effri_out, effrs_inout, ncl_out, errmsg, errflg)
 
       integer,            intent(in)    :: ncol
       integer,            intent(in)    :: nlev
       real(kind_phys),    intent(in)    :: effrr_in(:,:)
-      real(kind_phys),    intent(in)    :: effrg_in(:,:)
+      real(kind_phys),    intent(in),optional    :: effrg_in(:,:)
+      real(kind_phys),    intent(in),optional    :: ncg_in(:,:)
+      real(kind_phys),    intent(out),optional   :: nci_out(:,:)
       real(kind_phys),    intent(inout) :: effrl_inout(:,:)
-      real(kind_phys),    intent(out)   :: effri_out(:,:)
+      real(kind_phys),    intent(out),optional   :: effri_out(:,:)
       real(8),intent(inout) :: effrs_inout(:,:)
-      logical,            intent(in)    :: has_graupel
+      real(kind_phys),    intent(out),optional   :: ncl_out(:,:)
       character(len=512), intent(out)   :: errmsg
       integer,            intent(out)   :: errflg
```

Comment:
I agree this is ugly, and will fix, but I believe the standard in this case is referring to the argument list, not the declarations within the procedure.

Reply:
Right but for instance, `errmsg` and `errflg` are not `optional` but show up in the argument list after `effrg_in`, `effri_out`, and others.

Reply:
This is something we've done consistently - always have errmsg and errflg at the very end. This is even more compelling if we think about the metadata order and the requirement (not sure if it's enforced in capgen, it isn't yet in prebuild) to have the same order as the argument list. Keeping those non-physics and purely framework related variables at the end would be nice.

Since nobody should be calling CCPP-compliant schemes directly - i.e. without ccpp-framework - we can rely on the arguments being listed as keywords?

Reply:
Or, alternatively, move the framework-mandated variables to the very beginning? An awful lot of changes in the current schemes in the physics, but doable.

Reply:
> Since nobody should be calling CCPP-compliant schemes directly - i.e. without ccpp-framework - we can rely on the arguments being listed as keywords?

At the very least, make this a bold part of the documentation about optional variables. Anyone creating a new scheme can put the variables in any order and anyone is free to use current schemes outside of the CCPP but we should at least document the consequences.

Reply:
I created issue https://github.com/NCAR/ccpp-framework/issues/544 for that.